marshmallow_dataclass/generic_resolver.py

```diff
@@ -0,0 +1,193 @@
+import copy
+import dataclasses
+import inspect
+import sys
+from typing import (
+    Dict,
+    Generic,
+    List,
+    Optional,
+    Tuple,
+    TypeVar,
+)
+
+import typing_extensions
+import typing_inspect
+
+if sys.version_info >= (3, 9):
+    from typing import Annotated
+else:
+    from typing_extensions import Annotated
+
+_U = TypeVar("_U")
+
+
+class UnboundTypeVarError(TypeError):
+    """TypeVar instance can not be resolved to a type spec.
+
+    This exception is raised when an unbound TypeVar is encountered.
+
+    """
+
+
+class InvalidStateError(Exception):
+    """Raised when an operation is performed on a future that is not
+    allowed in the current state.
+    """
+
+
+class _Future(Generic[_U]):
+    """The _Future class allows deferred access to a result that is not
+    yet available.
+    """
+
+    _done: bool
+    _result: _U
+
+    def __init__(self) -> None:
+        self._done = False
+
+    def done(self) -> bool:
+        """Return ``True`` if the value is available"""
+        return self._done
+
+    def result(self) -> _U:
+        """Return the deferred value.
+
+        Raises ``InvalidStateError`` if the value has not been set.
+        """
+        if self.done():
+            return self._result
+        raise InvalidStateError("result has not been set")
+
+    def set_result(self, result: _U) -> None:
+        if self.done():
+            raise InvalidStateError("result has already been set")
+        self._result = result
+        self._done = True
+
+
+def is_generic_alias(clazz: type) -> bool:
+    """
+    Check if given class is a generic alias of a class is
+    defined as `class A(Generic[T])`, this method will return true if `A[int]` is passed
+    """
+    is_generic = is_generic_type(clazz)
+    type_arguments = typing_extensions.get_args(clazz)
+    return is_generic and len(type_arguments) > 0
+
+
+def is_generic_type(clazz: type) -> bool:
+    """
+    typing_inspect.is_generic_type explicitly ignores Union, Tuple, Callable, ClassVar
```

Comment:
I'm not sure what change is being asked for here? A name change, or change in logic maybe?

I can't think of any usecase where we'd need to handle ClassVar and Callable to generate a marshmallow object, so I can add exclusion of them. 
But Tuple and Union are standard marshmallow fields, so that makes more sense to support.

Reply:
I'm not sure what change is best here, either.  Clearly, the current function behavior does not match its name. (So, that is confusing.)

I'm not sure this function is necessary.  It seems most uses of it are on objects that are already known (or assumed to be dataclasses), in which case all these edge cases are moot.  (The other place it is used is on what are potentially generic marshmallow `Field`s.)

Note that unions are special-cased in `_field_for_union_type` and tuples (as well as all the "well known" containers — dicts, lists, etc) are special-cased in `_field_for_generic_type`.

Which brings up: `_field_for_generic_type` should probably be renamed to `_field_for_container_type` (or some such), since it doesn't really have anything to do with the support for generics added by this PR.

Reply:
I've renamed the relevant functions and variables to `container`

I've dug a little deeper and we only need the `is_generic_type` to see if we need to resolve typevars. So I'll rename it to  `may_contain_typevars(clazz)` and only use it where applicable. And use `typing_inspect.is_generic_type` elsewhere

Reply:
@dairiki I'm going to mark this thread as resolved. Re-open if you disagree.